File: chess_coach/ingest.py
"""Chess.com ingestion.

Pulls a player's public games from the chess.com Published-Data API and parses
them into the internal `Game` model.

API shape (all public, no auth):
    GET /pub/player/{username}/games/archives
        -> {"archives": ["https://api.chess.com/pub/player/{u}/games/2024/01", ...]}
    GET /pub/player/{username}/games/{YYYY}/{MM}
        -> {"games": [ {url, pgn, time_control, time_class, rated, rules,
                        end_time, eco?, white:{...}, black:{...}}, ... ]}

Notes that bite people:
  * chess.com now REQUIRES a descriptive User-Agent or returns 403. Set a real
    contact string.
  * Endpoints are Cloudflare-cached and rate-limited; on 429 we honor
    Retry-After and back off.
  * Games include variants (chess960, bughouse, ...). We keep only standard
    `rules == "chess"` by default so the eval/tagging layers can assume normal
    chess.
"""

from __future__ import annotations

import io
import time
from typing import Iterable, Iterator, Optional

import chess
import chess.pgn
import requests

from .models import (
    Color, Game, Player, Ply, epoch_to_dt,
)

API_BASE = "https://api.chess.com"
DEFAULT_UA = "chess-coach/0.1 (https://github.com/yourname/chess-coach; contact@example.com)"


class ChessComError(RuntimeError):
    pass


class PlayerNotFound(ChessComError):
    pass


class ChessComClient:
    def __init__(
        self,
        user_agent: str = DEFAULT_UA,
        *,
        timeout: float = 20.0,
        max_retries: int = 4,
        session: Optional[requests.Session] = None,
    ):
        self.timeout = timeout
        self.max_retries = max_retries
        self.session = session or requests.Session()
        self.session.headers.update({
            "User-Agent": user_agent,
            "Accept": "application/json",
        })

    # ---------------- low-level HTTP ----------------
    def _get(self, url: str) -> requests.Response:
        backoff = 1.0
        for attempt in range(self.max_retries + 1):
            resp = self.session.get(url, timeout=self.timeout)
            if resp.status_code == 200:
                return resp
            if resp.status_code == 404:
                raise PlayerNotFound(f"404 Not Found: {url}")
            if resp.status_code == 429 and attempt < self.max_retries:
                wait = float(resp.headers.get("Retry-After", backoff))
                time.sleep(wait)
                backoff = min(backoff * 2, 30)
                continue
            if resp.status_code >= 500 and attempt < self.max_retries:
                time.sleep(backoff)
                backoff = min(backoff * 2, 30)
                continue
            raise ChessComError(f"{resp.status_code} for {url}: {resp.text[:200]}")
        raise ChessComError(f"Exhausted retries for {url}")

    # ---------------- archives ----------------
    def list_archive_urls(self, username: str) -> list[str]:
        """Monthly archive URLs, oldest first."""
        url = f"{API_BASE}/pub/player/{username.lower()}/games/archives"
        data = self._get(url).json()
        return data.get("archives", [])

    def _fetch_archive(self, archive_url: str) -> list[dict]:
        return self._get(archive_url).json().get("games", [])

    # ---------------- public: iterate games ----------------
    def iter_games(
        self,
        username: str,
        *,
        rules: Optional[set[str]] = frozenset({"chess"}),
        time_classes: Optional[set[str]] = None,
        rated_only: bool = False,
        max_games: Optional[int] = None,
        newest_first: bool = True,
    ) -> Iterator[Game]:
        """Yield parsed `Game` objects for a username.

        rules: keep only these rule sets (default: standard chess). None = all.
        time_classes: e.g. {"blitz","rapid"}. None = all.
        max_games: stop after N (useful for a quick first pass).
        newest_first: process most recent archives first.
        """
        username = username.lower()
        archives = self.list_archive_urls(username)
        if newest_first:
            archives = list(reversed(archives))

        count = 0
        for archive_url in archives:
            raw_games = self._fetch_archive(archive_url)
            if newest_first:
                raw_games = list(reversed(raw_games))
            for raw in raw_games:
                if rules is not None and raw.get("rules") not in rules:
                    continue
                if time_classes is not None and raw.get("time_class") not in time_classes:
                    continue
                if rated_only and not raw.get("rated", False):
                    continue
                game = parse_game(raw, perspective_username=username)
                if game is None:
                    continue
                yield game
                count += 1
                if max_games is not None and count >= max_games:
                    return

    def fetch_games(self, username: str, **kwargs) -> list[Game]:
        return list(self.iter_games(username, **kwargs))


# ---------------- parsing ----------------
def _player_from_raw(raw_side: dict) -> Player:
    return Player(
        username=(raw_side.get("username") or "").lower(),
        rating=raw_side.get("rating"),
        result_raw=raw_side.get("result"),
    )


def parse_game(raw: dict, perspective_username: str) -> Optional[Game]:
    """Turn one chess.com game dict into an internal Game.

    Returns None for games we can't/shouldn't parse (no PGN, unfinished daily
    game with empty movetext, etc.).
    """
    pgn_text = raw.get("pgn")
    if not pgn_text:
        return None

    white = _player_from_raw(raw.get("white", {}))
    black = _player_from_raw(raw.get("black", {}))

    pu = perspective_username.lower()
    if pu == white.username:
        perspective = Color.WHITE
    elif pu == black.username:
        perspective = Color.BLACK
    else:
        # Username didn't match either side (rare: name change). Default white.
        perspective = Color.WHITE

    moves = parse_moves(pgn_text)

    return Game(
        game_id=raw.get("uuid") or raw.get("url", ""),
        url=raw.get("url", ""),
        played_at=epoch_to_dt(raw.get("end_time")),
        time_class=raw.get("time_class", ""),
        time_control=raw.get("time_control", ""),
        rated=bool(raw.get("rated", False)),
        rules=raw.get("rules", "chess"),
        eco=raw.get("eco"),
        white=white,
        black=black,
        perspective=perspective,
        moves=moves,
        pgn=pgn_text,
    )


def parse_moves(pgn_text: str) -> list[Ply]:
    """Replay a PGN into a list of Ply objects with FEN before/after each move."""
    game = chess.pgn.read_game(io.StringIO(pgn_text))
    if game is None:
        return []

    board = game.board()
    plies: list[Ply] = []
    ply_no = 0
    for node in game.mainline():
        move = node.move
        fen_before = board.fen()
        color = Color.WHITE if board.turn == chess.WHITE else Color.BLACK
        move_number = board.fullmove_number
        san = board.san(move)
        board.push(move)
        ply_no += 1
        plies.append(Ply(
            ply_number=ply_no,
            move_number=move_number,
            color=color,
            san=san,
            uci=move.uci(),
            fen_before=fen_before,
            fen_after=board.fen(),
            clock_seconds=_clock_from_node(node),
        ))
    return plies


def _clock_from_node(node) -> Optional[float]:
    """Extract [%clk H:MM:SS] from a PGN node comment, in seconds."""
    try:
        clk = node.clock()  # python-chess >= 1.0 parses %clk
    except Exception:
        return None
    return clk

File: app/src/engine/ingest.ts
// chess.com ingestion — TS port of chess_coach/ingest.py (uses chess.js).
import { Chess } from 'chess.js';
import type { Color, Game, Player, Ply } from './model';

const API_BASE = 'https://api.chess.com';
export const DEFAULT_UA = 'mangus/0.1 (https://github.com/walkap123/mangus; contact@example.com)';

export class PlayerNotFound extends Error {}
export class ChessComError extends Error {}

function parseClk(comment: string): number | null {
  const m = comment.match(/\[%clk (\d+):(\d+):(\d+(?:\.\d+)?)\]/);
  if (!m) return null;
  return parseInt(m[1], 10) * 3600 + parseInt(m[2], 10) * 60 + parseFloat(m[3]);
}

export function parseMoves(pgn: string): Ply[] {
  const chess = new Chess();
  try {
    chess.loadPgn(pgn);
  } catch {
    return [];
  }
  const clkByFen = new Map<string, number | null>();
  for (const c of chess.getComments()) clkByFen.set(c.fen, parseClk(c.comment));

  const plies: Ply[] = [];
  let plyNo = 0;
  for (const mv of chess.history({ verbose: true })) {
    plyNo++;
    plies.push({
      plyNumber: plyNo,
      moveNumber: parseInt(mv.before.split(' ')[5], 10), // fullmove no. before the move
      color: mv.color === 'w' ? 'white' : 'black',
      san: mv.san,
      uci: mv.lan,
      fenBefore: mv.before,
      fenAfter: mv.after,
      clockSeconds: clkByFen.get(mv.after) ?? null,
    });
  }
  return plies;
}

function playerFromRaw(side: any): Player {
  return {
    username: (side.username || '').toLowerCase(),
    rating: side.rating ?? null,
    resultRaw: side.result ?? null,
  };
}

export function parseGame(raw: any, perspectiveUsername: string): Game | null {
  const pgn = raw.pgn;
  if (!pgn) return null;
  const white = playerFromRaw(raw.white || {});
  const black = playerFromRaw(raw.black || {});
  const pu = perspectiveUsername.toLowerCase();
  const perspective: Color = pu === white.username ? 'white' : pu === black.username ? 'black' : 'white';
  return {
    gameId: raw.uuid || raw.url || '',
    url: raw.url || '',
    playedAt: raw.end_time ?? null,
    timeClass: raw.time_class || '',
    timeControl: raw.time_control || '',
    rated: !!raw.rated,
    rules: raw.rules || 'chess',
    eco: raw.eco ?? null,
    white,
    black,
    perspective,
    moves: parseMoves(pgn),
    pgn,
  };
}

export interface IterOpts {
  rules?: Set<string> | null;
  timeClasses?: Set<string> | null;
  ratedOnly?: boolean;
  maxGames?: number | null;
  newestFirst?: boolean;
}

export class ChessComClient {
  constructor(private userAgent: string = DEFAULT_UA, private maxRetries = 4) {}

  private async get(url: string): Promise<Response> {
    let backoff = 1000;
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      const resp = await fetch(url, {
        headers: { 'User-Agent': this.userAgent, Accept: 'application/json' },
      });
      if (resp.status === 200) return resp;
      if (resp.status === 404) throw new PlayerNotFound(`404 Not Found: ${url}`);
      if ((resp.status === 429 || resp.status >= 500) && attempt < this.maxRetries) {
        const ra = parseFloat(resp.headers.get('Retry-After') || '') * 1000;
        await new Promise((r) => setTimeout(r, Number.isFinite(ra) && ra > 0 ? ra : backoff));
        backoff = Math.min(backoff * 2, 30000);
        continue;
      }
      throw new ChessComError(`${resp.status} for ${url}`);
    }
    throw new ChessComError(`Exhausted retries for ${url}`);
  }

  async listArchiveUrls(username: string): Promise<string[]> {
    const url = `${API_BASE}/pub/player/${username.toLowerCase()}/games/archives`;
    return (await (await this.get(url)).json()).archives || [];
  }

  private async fetchArchive(url: string): Promise<any[]> {
    return (await (await this.get(url)).json()).games || [];
  }

  async fetchGames(username: string, opts: IterOpts = {}): Promise<Game[]> {
    username = username.toLowerCase();
    const newestFirst = opts.newestFirst ?? true;
    const rules = opts.rules === undefined ? new Set(['chess']) : opts.rules;
    let archives = await this.listArchiveUrls(username);
    if (newestFirst) archives = archives.slice().reverse();

    const out: Game[] = [];
    for (const archiveUrl of archives) {
      let raws = await this.fetchArchive(archiveUrl);
      if (newestFirst) raws = raws.slice().reverse();
      for (const raw of raws) {
        if (rules && !rules.has(raw.rules)) continue;
        if (opts.timeClasses && !opts.timeClasses.has(raw.time_class)) continue;
        if (opts.ratedOnly && !raw.rated) continue;
        const g = parseGame(raw, username);
        if (!g) continue;
        out.push(g);
        if (opts.maxGames != null && out.length >= opts.maxGames) return out;
      }
    }
    return out;
  }
}
